# Demo -  Sakila Star Schema & ETL  

All the database tables in this demo are based on public database samples and transformations
- `Sakila` is a sample database created by `MySql` [Link](https://dev.mysql.com/doc/sakila/en/sakila-structure.html)
- The postgresql version of it is called `Pagila` [Link](https://github.com/devrimgunduz/pagila)
- The facts and dimension tables design

In [ ]:
# In case your env does not have the needed libraries
# !pip install psycopg2-binary
# !pip install ipython-sql
# !pip install pandas

# STEP0: Using ipython-sql

- Load ipython-sql: `%load_ext sql`
    - install ipython-sql 0.4.0 `pip install ipython-sql`

- To execute SQL queries you write one of the following atop of your cell: 
    - `%sql`
        - For a one-liner SQL query
        - You can access a python var using `$`    
    - `%%sql`
        - For a multi-line SQL query
        - You can **NOT** access a python var using `$`


- Running a connection string like:
`postgresql://postgres:postgres@db:5432/pagila` connects to the database


# STEP1 : Connect to the local database where Pagila is loaded

##  1.1 Create the pagila db and fill it with data
- Adding `"!"` at the beginning of a jupyter cell runs a command in a shell, i.e. we are not running python code but we are running the `createdb` and `psql` postgresql commmand-line utilities

In [3]:
# !PGPASSWORD=admin createdb -h 127.0.0.1 -U postgres pagila
# !PGPASSWORD=admin psql -q -h 127.0.0.1 -U postgres -d pagila -f Data/pagila-schema.sql
# !PGPASSWORD=admin psql -q -h 127.0.0.1 -U postgres -d pagila -f Data/pagila-data.sql

## 1.2 Connect to the newly created db

In [4]:
%load_ext sql

In [5]:
DB_ENDPOINT = "127.0.0.1"
DB = 'pagila'
DB_USER = 'postgres'
DB_PASSWORD = 'admin'
DB_PORT = '5432'

# postgresql://username:password@host:port/database
conn_string = "postgresql://{}:{}@{}:{}/{}" \
                        .format(DB_USER, DB_PASSWORD, DB_ENDPOINT, DB_PORT, DB)

print(conn_string)


postgresql://postgres:admin@127.0.0.1:5432/pagila


In [6]:
%sql $conn_string

/Users/kholed/anaconda3/lib/python3.7/site-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


# STEP2 : Explore the  3NF Schema

<img src="images/pagila-3nf.png" width="70%"/>

## 2.1 How much? What data sizes are we looking at?

In [20]:
nStores = %sql select count(*) from store;
nFilms = %sql select count(*) from film;
nCustomers = %sql select count(*) from customer;
nRentals = %sql select count(*) from rental;
nPayment = %sql select count(*) from payment;
nStaff = %sql select count(*) from staff;
nCity = %sql select count(*) from city;
nCountry = %sql select count(*) from country;

print("nFilms\t\t=", nFilms[0][0])
print("nCustomers\t=", nCustomers[0][0])
print("nRentals\t=", nRentals[0][0])
print("nPayment\t=", nPayment[0][0])
print("nStaff\t\t=", nStaff[0][0])
print("nStores\t\t=", nStores[0][0])
print("nCities\t\t=", nCity[0][0])
print("nCountry\t\t=", nCountry[0][0])

 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.
nFilms		= 1000
nCustomers	= 599
nRentals	= 16044
nPayment	= 16049
nStaff		= 2
nStores		= 2
nCities		= 600
nCountry		= 109


## 2.2 When? What time period are we talking about?

In [26]:
%%sql 
select min(rental_date) as start, max(rental_date) as end from rental;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.


[(datetime.datetime(2005, 5, 25, 4, 53, 30, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), datetime.datetime(2020, 2, 14, 22, 16, 3, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)))]

In [21]:
%%sql 
select min(payment_date) as start, max(payment_date) as end from payment;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
1 rows affected.


[(datetime.datetime(2020, 1, 25, 4, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), datetime.datetime(2020, 5, 14, 19, 44, 29, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)))]

## 2.3 Where? Where do events in this database occur?
TODO: Write a query that displays the number of addresses by district in the address table. Limit the table to the top 10 districts. Your results should match the table below.

In [25]:
%%sql
select district, sum(city_id) as n
from address
group by district
order by n desc
limit 10 ;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('Shandong', 3237),
 ('England', 2974),
 ('So Paulo', 2952),
 ('West Bengali', 2623),
 ('Buenos Aires', 2572),
 ('Uttar Pradesh', 2462),
 ('California', 2444),
 ('Southern Tagalog', 1931),
 ('Tamil Nadu', 1807),
 ('Hubei', 1790)]

In [23]:
%%sql
select district, count(1)
from address
group by district
order by count(1) desc
limit 10


 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('Buenos Aires', 10),
 ('Shandong', 9),
 ('California', 9),
 ('West Bengali', 9),
 ('Uttar Pradesh', 8),
 ('So Paulo', 8),
 ('England', 7),
 ('Maharashtra', 7),
 ('Southern Tagalog', 6),
 ('Gois', 5)]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>district</th>
        <th>n</th>
    </tr>
    <tr>
        <td>Buenos Aires</td>
        <td>10</td>
    </tr>
    <tr>
        <td>California</td>
        <td>9</td>
    </tr>
    <tr>
        <td>Shandong</td>
        <td>9</td>
    </tr>
    <tr>
        <td>West Bengali</td>
        <td>9</td>
    </tr>
    <tr>
        <td>So Paulo</td>
        <td>8</td>
    </tr>
    <tr>
        <td>Uttar Pradesh</td>
        <td>8</td>
    </tr>
    <tr>
        <td>Maharashtra</td>
        <td>7</td>
    </tr>
    <tr>
        <td>England</td>
        <td>7</td>
    </tr>
    <tr>
        <td>Southern Tagalog</td>
        <td>6</td>
    </tr>
    <tr>
        <td>Punjab</td>
        <td>5</td>
    </tr>
</tbody></table></div>

# STEP3: Perform some simple data analysis

### 3NF - Entity Relationship Diagram

<img src="images/pagila-3nf.png" width="70%"/>

## 3.1 Insight 1:   Top Grossing Movies 
- Payments amounts are in table `payment`
- Movies are in table `film`
- They are not directly linked, `payment` refers to a `rental`, `rental` refers to an `inventory` item and `inventory` item refers to a `film`
- `payment` &rarr; `rental` &rarr; `inventory` &rarr; `film`

### 3.1.1 Films

In [27]:
%%sql
select film_id, title, release_year, rental_rate, rating  from film limit 5;


 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 'ACADEMY DINOSAUR', 2006, Decimal('0.99'), 'PG'),
 (2, 'ACE GOLDFINGER', 2006, Decimal('4.99'), 'G'),
 (3, 'ADAPTATION HOLES', 2006, Decimal('2.99'), 'NC-17'),
 (4, 'AFFAIR PREJUDICE', 2006, Decimal('2.99'), 'G'),
 (5, 'AFRICAN EGG', 2006, Decimal('2.99'), 'G')]

### 3.1.2 Payments

In [28]:
%%sql
select * from payment limit 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(16050, 269, 2, 7, Decimal('1.99'), datetime.datetime(2020, 1, 25, 4, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (16051, 269, 1, 98, Decimal('0.99'), datetime.datetime(2020, 1, 25, 22, 16, 50, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (16052, 269, 2, 678, Decimal('6.99'), datetime.datetime(2020, 1, 29, 4, 44, 14, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (16053, 269, 2, 703, Decimal('0.99'), datetime.datetime(2020, 1, 29, 7, 58, 2, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (16054, 269, 1, 750, Decimal('4.99'), datetime.datetime(2020, 1, 29, 15, 10, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)))]

### 3.1.3 Inventory

In [29]:
%%sql
select * from inventory limit 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(1, 1, 1, datetime.datetime(2020, 2, 15, 17, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (2, 1, 1, datetime.datetime(2020, 2, 15, 17, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (3, 1, 1, datetime.datetime(2020, 2, 15, 17, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (4, 1, 1, datetime.datetime(2020, 2, 15, 17, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None))),
 (5, 1, 2, datetime.datetime(2020, 2, 15, 17, 9, 17, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)))]

### 3.1.4 Get the movie of every payment

In [30]:
%%sql
SELECT f.title, p.amount, p.payment_date, p.customer_id                                            
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
limit 5;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[('SWARM GOLD', Decimal('1.99'), datetime.datetime(2020, 1, 25, 4, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), 269),
 ('PACKER MADIGAN', Decimal('0.99'), datetime.datetime(2020, 1, 25, 22, 16, 50, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), 269),
 ('SOMETHING DUCK', Decimal('6.99'), datetime.datetime(2020, 1, 29, 4, 44, 14, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), 269),
 ('DRACULA CRYSTAL', Decimal('0.99'), datetime.datetime(2020, 1, 29, 7, 58, 2, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), 269),
 ('CLOSER BANG', Decimal('4.99'), datetime.datetime(2020, 1, 29, 15, 10, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), 269)]

### 3.1.5 sum movie rental revenue
TODO: Write a query that displays the amount of revenue from each title. Limit the results to the top 10 grossing titles. Your results should match the table below.

In [31]:
%%sql
SELECT f.title, sum(p.amount) as revenue                                           
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
group by f.title
order by revenue desc
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('TELEGRAPH VOYAGE', Decimal('231.73')),
 ('WIFE TURN', Decimal('223.69')),
 ('ZORRO ARK', Decimal('214.69')),
 ('GOODFELLAS SALUTE', Decimal('209.69')),
 ('SATURDAY LAMBS', Decimal('204.72')),
 ('TITANS JERK', Decimal('201.71')),
 ('TORQUE BOUND', Decimal('198.72')),
 ('HARRY IDAHO', Decimal('195.70')),
 ('INNOCENT USUAL', Decimal('191.74')),
 ('HUSTLER PARTY', Decimal('190.78'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>231.73</td>
    </tr>
    <tr>
        <td>WIFE TURN</td>
        <td>223.69</td>
    </tr>
    <tr>
        <td>ZORRO ARK</td>
        <td>214.69</td>
    </tr>
    <tr>
        <td>GOODFELLAS SALUTE</td>
        <td>209.69</td>
    </tr>
    <tr>
        <td>SATURDAY LAMBS</td>
        <td>204.72</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>201.71</td>
    </tr>
    <tr>
        <td>TORQUE BOUND</td>
        <td>198.72</td>
    </tr>
    <tr>
        <td>HARRY IDAHO</td>
        <td>195.70</td>
    </tr>
    <tr>
        <td>INNOCENT USUAL</td>
        <td>191.74</td>
    </tr>
    <tr>
        <td>HUSTLER PARTY</td>
        <td>190.78</td>
    </tr>
</tbody></table></div>

Congratulation, you have run up to step 3

## 3.2 Insight 2:   Top grossing cities 
- Payments amounts are in table `payment`
- Cities are in table `cities`
- `payment` &rarr; `customer` &rarr; `address` &rarr; `city`

### 3.2.1 Get the city of each payment

In [15]:
%%sql
SELECT p.customer_id, p.rental_id, p.amount, ci.city                            
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[(130, 1, Decimal('2.99'), 'guas Lindas de Gois'),
 (459, 2, Decimal('2.99'), 'Qomsheh'),
 (408, 3, Decimal('3.99'), 'Jaffna'),
 (333, 4, Decimal('4.99'), 'Baku'),
 (222, 5, Decimal('6.99'), 'Jaroslavl'),
 (549, 6, Decimal('0.99'), 'Santiago de Compostela'),
 (269, 7, Decimal('1.99'), 'Salinas'),
 (239, 8, Decimal('4.99'), 'Ciomas'),
 (126, 9, Decimal('4.99'), 'Po'),
 (399, 10, Decimal('5.99'), 'Okara')]

### 3.2.2 Top grossing cities
TODO: Write a query that returns the total amount of revenue by city as measured by the `amount` variable in the `payment` table. Limit the results to the top 10 cities. Your result should match the table below.

In [16]:
%%sql
SELECT ci.city,sum(p.amount) as revenue
FROM payment p
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
group by ci.city
order by revenue desc
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('Cape Coral', Decimal('221.55')),
 ('Saint-Denis', Decimal('216.54')),
 ('Aurora', Decimal('198.50')),
 ('Molodetno', Decimal('195.58')),
 ('Santa Brbara dOeste', Decimal('194.61')),
 ('Apeldoorn', Decimal('194.61')),
 ('Qomsheh', Decimal('186.62')),
 ('London', Decimal('180.52')),
 ('Ourense (Orense)', Decimal('177.60')),
 ('Bijapur', Decimal('175.61'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>city</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>Cape Coral</td>
        <td>221.55</td>
    </tr>
    <tr>
        <td>Saint-Denis</td>
        <td>216.54</td>
    </tr>
    <tr>
        <td>Aurora</td>
        <td>198.50</td>
    </tr>
    <tr>
        <td>Molodetno</td>
        <td>195.58</td>
    </tr>
    <tr>
        <td>Apeldoorn</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Santa Brbara dOeste</td>
        <td>194.61</td>
    </tr>
    <tr>
        <td>Qomsheh</td>
        <td>186.62</td>
    </tr>
    <tr>
        <td>London</td>
        <td>180.52</td>
    </tr>
    <tr>
        <td>Ourense (Orense)</td>
        <td>177.60</td>
    </tr>
    <tr>
        <td>Bijapur</td>
        <td>175.61</td>
    </tr>
</tbody></table></div>

## 3.3 Insight 3 : Revenue of a movie by customer city and by month 

### 3.3.1 Total revenue by month

In [17]:
%%sql
SELECT sum(p.amount) as revenue, EXTRACT(month FROM p.payment_date) as month
from payment p
group by month
order by revenue desc
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
5 rows affected.


[(Decimal('27226.52'), Decimal('4')),
 (Decimal('23886.56'), Decimal('3')),
 (Decimal('9745.61'), Decimal('2')),
 (Decimal('4710.70'), Decimal('1')),
 (Decimal('1847.12'), Decimal('5'))]

### 3.3.2 Each movie by customer city and by month (data cube)

In [18]:
%%sql
SELECT f.title, p.amount, p.customer_id, ci.city, p.payment_date,EXTRACT(month FROM p.payment_date) as month
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
order by p.payment_date
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('BLANKET BEVERLY', Decimal('2.99'), 130, 'guas Lindas de Gois', datetime.datetime(2020, 1, 25, 4, 21, 56, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('FREAKY POCUS', Decimal('2.99'), 459, 'Qomsheh', datetime.datetime(2020, 1, 25, 4, 22, 59, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('GRADUATE LORD', Decimal('3.99'), 408, 'Jaffna', datetime.datetime(2020, 1, 25, 4, 32, 5, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('LOVE SUICIDES', Decimal('4.99'), 333, 'Baku', datetime.datetime(2020, 1, 25, 4, 33, 7, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('IDOLS SNATCHERS', Decimal('6.99'), 222, 'Jaroslavl', datetime.datetime(2020, 1, 25, 4, 33, 47, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('MYSTIC TRUMAN', Decimal('0.99'), 549, 'Santiago de Compostela', datetime.datetime(2020, 1, 25, 4, 36, 33, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('SWARM GOLD', Decimal('1.99'), 269, 'Salinas', datetime.datetime(2020, 1, 25, 4, 40, 19, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('LAWLESS VISION', Decimal('4.99'), 239, 'Ciomas', datetime.datetime(2020, 1, 25, 5, 0, 12, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('MATRIX SNOWMAN', Decimal('4.99'), 126, 'Po', datetime.datetime(2020, 1, 25, 5, 29, 6, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1')),
 ('HANGING DEEP', Decimal('5.99'), 399, 'Okara', datetime.datetime(2020, 1, 25, 5, 30, 47, 996577, tzinfo=psycopg2.tz.FixedOffsetTimezone(offset=420, name=None)), Decimal('1'))]

### 3.3.3 Sum of revenue of each movie by customer city and by month

TODO: Write a query that returns the total amount of revenue for each movie by customer city and by month. Limit the results to the top 10 movies. Your result should match the table below.

In [19]:
%%sql
SELECT f.title, ci.city, EXTRACT(month FROM p.payment_date) as month,sum(p.amount) as revenue
FROM payment p
JOIN rental r    ON ( p.rental_id = r.rental_id )
JOIN inventory i ON ( r.inventory_id = i.inventory_id )
JOIN film f ON ( i.film_id = f.film_id)
JOIN customer c  ON ( p.customer_id = c.customer_id )
JOIN address a ON ( c.address_id = a.address_id )
JOIN city ci ON ( a.city_id = ci.city_id )
group by f.title, ci.city, month
order by revenue desc
limit 10;

 * postgresql://postgres:***@127.0.0.1:5432/pagila
10 rows affected.


[('EAGLES PANKY', 'Datong', Decimal('3'), Decimal('16.98')),
 ('CARIBBEAN LIBERTY', 'Ibirit', Decimal('4'), Decimal('16.98')),
 ('LUST LOCK', 'Coatzacoalcos', Decimal('4'), Decimal('15.98')),
 ('INNOCENT USUAL', 'Valparai', Decimal('2'), Decimal('13.98')),
 ('BREAKFAST GOLDFINGER', 'Ife', Decimal('4'), Decimal('13.98')),
 ('VIRGIN DAISY', 'Cam Ranh', Decimal('4'), Decimal('12.98')),
 ('SLEEPLESS MONSOON', 'Fuyu', Decimal('4'), Decimal('12.98')),
 ('TITANIC BOONDOCK', 'Sumqayit', Decimal('4'), Decimal('12.98')),
 ('MIDSUMMER GROUNDHOG', 'Szkesfehrvr', Decimal('4'), Decimal('11.99')),
 ('SCORPION APOLLO', 'Osmaniye', Decimal('4'), Decimal('11.99'))]

<div class="p-Widget jp-RenderedHTMLCommon jp-RenderedHTML jp-mod-trusted jp-OutputArea-output jp-OutputArea-executeResult" data-mime-type="text/html"><table>
    <tbody><tr>
        <th>title</th>
        <th>city</th>
        <th>month</th>
        <th>revenue</th>
    </tr>
    <tr>
        <td>SHOW LORD</td>
        <td>Mannheim</td>
        <td>1.0</td>
        <td>11.99</td>
    </tr>
    <tr>
        <td>AMERICAN CIRCUS</td>
        <td>Callao</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>CASUALTIES ENCINO</td>
        <td>Warren</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>TELEGRAPH VOYAGE</td>
        <td>Naala-Porto</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>KISSING DOLLS</td>
        <td>Toulon</td>
        <td>1.0</td>
        <td>10.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Bergamo</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>TITANS JERK</td>
        <td>Kimberley</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>DARKO DORADO</td>
        <td>Bhilwara</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>SUNRISE LEAGUE</td>
        <td>Nagareyama</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
    <tr>
        <td>MILLION ACE</td>
        <td>Gaziantep</td>
        <td>1.0</td>
        <td>9.99</td>
    </tr>
</tbody></table></div>